File: src/cli_server/weather_py_server/weather.py
"""
Access National Weather Service data for US locations, including forecasts and alerts.
"""

import sys
import logging

import asyncio

import inspect
from typing import Any, Annotated, Callable
from pydantic import Field

import httpx

from weather_helpers import make_nws_request, format_alert

LOG_FORMAT = "\033[32m%(levelname)s\033[0m:    %(message)s"
logging.basicConfig(level=logging.INFO, format=LOG_FORMAT)
logger = logging.getLogger("weather")


# Constants
NWS_API_BASE = "https://api.weather.gov"

# Implementing tool execution
async def get_alerts(
    state: Annotated[str, Field(description="Two-letter US state code (e.g. CA, NY)")]
) -> str:
    """Get weather alerts for a US state.
    
    Args:
        state: Two-letter US state code (e.g. CA, NY)
    """
    url = f"{NWS_API_BASE}/alerts/active?area={state.upper()}"
    data = await make_nws_request(url)
    if not data or "features" not in data:
        return "Unable to fetch alerts or no alerts found."
    
    features = data["features"]
    if not features:
        return "No active alerts for this state."
    
    alerts = [format_alert(feature) for feature in features]
    return "\n---\n".join(alerts)


async def get_forecast(
    latitude: Annotated[float, Field(description="The latitude of the location.")],
    longitude: Annotated[float, Field(description="The longitude of the location.")]
) -> str:
    """Get weather forecast for a location.

    Args:
        latitude: The latitude of the location.
        longitude: The longitude of the location.
    """

    # First get the forecast grid endpoint
    points_url = f"{NWS_API_BASE}/points/{latitude},{longitude}"
    points_data = await make_nws_request(points_url)

    if not points_data:
        return "Unable to fetch forecast data for this location."

    # Get the forecast URL from the points response
    forecast_url = points_data["properties"]["forecast"]
    forecast_data = await make_nws_request(forecast_url)

    if not forecast_data:
        return "Unable to fetch detailed forecast."
    
    # Format the periods into a readable forecast string
    periods = forecast_data["properties"]["periods"]
    forecasts = []
    for period in periods[:5]: # Only show next 5 periods
        forecast = f"""
{period["name"]}:
Temperature: {period["temperature"]}°{period["temperatureUnit"]}
Wind: {period["windSpeed"]} {period["windDirection"]}
Forecast: {period["detailedForecast"]}
"""
        forecasts.append(forecast)
    
    return "\n---\n".join(forecasts)



# def main():
#     result = asyncio.run(get_forecast(34.0522, -118.2437))
#     print(result[:200])
#     logging.info("Example forecast retrieval complete.")

# if __name__ == "__main__":
#     main()

File: src_ts/cli_server/weather_ts_server/weather.ts
/**
 * Access National Weather Service data for US locations, including forecasts and alerts.
 */

import { makeNwsRequest, formatAlert } from "./weather_helpers";

const NWS_API_BASE = "https://api.weather.gov";

export async function getAlerts(state: string): Promise<string> {
  const url = `${NWS_API_BASE}/alerts/active?area=${state.toUpperCase()}`;
  const data = await makeNwsRequest(url);
  if (!data || !data.features) return "Unable to fetch alerts or no alerts found.";
  const features = data.features as any[];
  if (!features.length) return "No active alerts for this state.";
  return features.map(formatAlert).join("\n---\n");
}

export async function getForecast(latitude: number, longitude: number): Promise<string> {
  const lat = Math.round(latitude * 10000) / 10000;
  const lon = Math.round(longitude * 10000) / 10000;

  const pointsUrl = `${NWS_API_BASE}/points/${lat},${lon}`;
  const pointsData = await makeNwsRequest(pointsUrl);
  if (!pointsData) return "Unable to fetch forecast data for this location.";

  const forecastUrl = pointsData?.properties?.forecast;
  if (!forecastUrl) return "Unable to fetch detailed forecast.";

  const forecastData = await makeNwsRequest(forecastUrl);
  if (!forecastData) return "Unable to fetch detailed forecast.";

  const periods: any[] = forecastData?.properties?.periods ?? [];
  const chunks = periods.slice(0, 5).map((period) => {
    return `\n${period.name}:\nTemperature: ${period.temperature}°${period.temperatureUnit}\nWind: ${period.windSpeed} ${period.windDirection}\nForecast: ${period.detailedForecast}\n`;
  });
  return chunks.join("\n---\n");
}

